File: logic/classes/Rule.py
from classes.Algorithm import Algorithm
from classes.Fact import Fact


def extract_new_facts(string):
	remove_list = ['!', '+', '|', '^', '(', ')', '\n']
	for char in remove_list:
		if char in string:
			string = string.replace(char, ' ')

	return string.split(' ')


class Rule(object):
	def __init__(self, rule, parent):
		super(Rule, self).__init__()
		self.condition = ''
		self.implication = ''

		self.parent = parent
		self.rule = rule
		self.algorithm = Algorithm(self.rule)
		self.process_rule()

	def process_rule(self):
		self.condition = self.algorithm.shunting_yard(self.algorithm.condition)
		self.implication = self.algorithm.shunting_yard(self.algorithm.implication)

		facts = extract_new_facts(self.algorithm.condition)
		self.collect_new_facts(facts, False)
		facts = extract_new_facts(self.algorithm.implication)
		self.collect_new_facts(facts, True)

	def collect_new_facts(self, facts, implication):
		for fact in facts:
			if len(fact) == 1:
				if not fact in self.parent.facts:
					self.parent.facts[fact] = Fact(fact, self.parent)
				if implication:
					# print fact, self.condition
					self.parent.facts[fact].dependencies.append(self)

File: logic/classes/Rule.js
const { Algorithm } = require('./Alghorithm');
const { Fact } = require('./Fact');

const extractNewFacts = (str) => {
    const removeList = ['!', '+', '|', '^', '(', ')', '\n'];
    for (let char of removeList) {
        if (str.includes(char)) {
            str = str.replace(char, ' ');
        }
    }
    return str.split(' ');
};


class Rule {
    constructor(rule, parent) {
        this.condition = '';
		this.implication = '';

		this.parent = parent;
		this.rule = rule;
		this.algorithm = new Algorithm(this.rule);
		this.processRule();
    }

    processRule() {
        this.condition = this.algorithm.shuntingYard(this.algorithm.condition);
		this.implication = this.algorithm.shuntingYard(this.algorithm.implication);

		let facts = extractNewFacts(this.algorithm.condition);
		this.collectNewFacts(facts, false);
		facts = extractNewFacts(this.algorithm.implication);
		this.collectNewFacts(facts, true);
    }

    collectNewFacts(facts, implication) {
        for (let fact of facts) {
            if (fact.length === 1) {
                if (!this.parent.facts.includes(fact)) {
                    this.parent.facts[fact] = new Fact(fact, this.parent);
                }
                if (implication) {
                    this.parent.facts[fact].dependencies.push(this);
                }
            }
        }
    }
}

module.exports = {
    extractNewFacts,
    Rule
};



